File: haskell/problem_104.hs
-- The Fibonacci sequence is defined by the recurrence relation:
-- 
-- Fn = Fn1 + Fn2, where F1 = 1 and F2 = 1.
-- 
-- It turns out that F541, which contains 113 digits, is the first Fibonacci
-- number for which the last nine digits are 1-9 pandigital (contain all the
-- digits 1 to 9, but not necessarily in order). And F2749, which contains 575
-- digits, is the first Fibonacci number for which the first nine digits are
-- 1-9 pandigital.
-- 
-- Given that Fk is the first Fibonacci number for which the first nine digits
-- AND the last nine digits are 1-9 pandigital, find k.
--
-- http://projecteuler.net/problem=104

module Main where

import Data.List ((\\), sort)

import MillerRabin

fibs :: [Integer]
fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

pandigital :: String -> Bool
pandigital s = (sort s) == ['1'..'9']

first_nine_pandigital :: Integer -> Bool
first_nine_pandigital n = 
  let n' = show n
  in pandigital (take 9 n')

last_nine_pandigital :: Integer -> Bool
last_nine_pandigital n = 
  let n' = reverse $ show n
  in pandigital (take 9 n')

main :: IO ()
main = do print $ length $ takeUntil (\f -> (first_nine_pandigital f) && (last_nine_pandigital f)) fibs

File: ruby/problem_104.rb
# The Fibonacci sequence is defined by the recurrence relation:
# 
# Fn = Fn1 + Fn2, where F1 = 1 and F2 = 1.
# 
# It turns out that F541, which contains 113 digits, is the first Fibonacci
# number for which the last nine digits are 1-9 pandigital (contain all the
# digits 1 to 9, but not necessarily in order). And F2749, which contains 575
# digits, is the first Fibonacci number for which the first nine digits are 1-9
# pandigital.
# 
# Given that Fk is the first Fibonacci number for which the first nine digits
# AND the last nine digits are 1-9 pandigital, find k.
#
# http://projecteuler.net/problem=104

class Fibonacci
  attr_accessor :i

  def initialize
    @fibonaccis = [0, 1]
  end

  def next
    @i ||= 0
    @i += 1
    @fibonaccis[@i] ||= @fibonaccis[@i - 2] + @fibonaccis[@i - 1]
  end
end

def pandigital?(str)
  (%w{1 2 3 4 5 6 7 8 9} - str.split(//)) == []
end

fibonacci = Fibonacci.new
f = 1
until pandigital?(f.to_s[0..9]) and pandigital?(f.to_s[-9..-1])
  f = fibonacci.next
  puts "f_#{fibonacci.i} = #{f}"
end
p f
p fibonacci.i
